File: cli/ielts/srs.py
"""SM-2 間隔重複演算法。

刻意寫成無副作用的純函式（不碰資料庫、不讀時間），方便測試與之後替換成
其他排程器（例如 FSRS）。呼叫端負責把結果寫回 srs_state。
"""

from __future__ import annotations

import random
from dataclasses import replace
from datetime import date, timedelta

from .models import SrsState, now_iso

AGAIN, HARD, GOOD, EASY = 1, 2, 3, 4
RATINGS = (AGAIN, HARD, GOOD, EASY)

RATING_LABELS = {
    AGAIN: "Again 忘了",
    HARD: "Hard 有點吃力",
    GOOD: "Good 想得起來",
    EASY: "Easy 太簡單",
}

RATING_SHORT = {AGAIN: "Again", HARD: "Hard", GOOD: "Good", EASY: "Easy"}

MIN_EASE = 1.3
MAX_EASE = 3.0
DEFAULT_EASE = 2.5
MAX_INTERVAL_DAYS = 365 * 2

#: 首次答對 1 天後再見；第二次答對跳到 6 天（SM-2 經典設定）。
FIRST_INTERVAL = 1.0
SECOND_INTERVAL = 6.0

#: 忘記後不是砍回 0，而是回到 1 天，並保留一點既有熟悉度。
LAPSE_INTERVAL = 1.0

HARD_MULTIPLIER = 1.2
EASY_BONUS = 1.3

EASE_DELTA = {AGAIN: -0.20, HARD: -0.15, GOOD: 0.0, EASY: 0.15}

#: 間隔 >= 這個天數才套用抖動，避免大量卡片集中在同一天到期。
JITTER_MIN_INTERVAL = 4.0
JITTER_RATIO = 0.05


def clamp_ease(value: float) -> float:
    return max(MIN_EASE, min(MAX_EASE, round(value, 4)))


def validate_rating(rating: int) -> int:
    if rating not in RATINGS:
        raise ValueError(f"評分必須是 1-4，收到 {rating!r}")
    return rating


def next_interval(state: SrsState, rating: int) -> float:
    """依評分算出新的間隔（天）。不含抖動。"""
    validate_rating(rating)
    interval = max(0.0, float(state.interval))
    ease = clamp_ease(state.ease_factor)

    if rating == AGAIN:
        return LAPSE_INTERVAL

    if state.review_count == 0 or interval <= 0:
        # 第一次答對
        return FIRST_INTERVAL if rating != EASY else SECOND_INTERVAL

    if interval < SECOND_INTERVAL and state.review_count == 1:
        # 第二次答對，進入 6 天
        base = SECOND_INTERVAL
        if rating == HARD:
            base = max(FIRST_INTERVAL + 1, interval * HARD_MULTIPLIER)
        elif rating == EASY:
            base = SECOND_INTERVAL * EASY_BONUS
        return base

    if rating == HARD:
        return interval * HARD_MULTIPLIER
    if rating == GOOD:
        return interval * ease
    return interval * ease * EASY_BONUS


def apply_jitter(interval: float, rng: random.Random | None = None) -> float:
    """對較長的間隔加 ±5% 隨機抖動，讓到期日自然分散。"""
    if interval < JITTER_MIN_INTERVAL:
        return interval
    r = rng or random
    factor = 1.0 + r.uniform(-JITTER_RATIO, JITTER_RATIO)
    return interval * factor


def schedule(
    state: SrsState,
    rating: int,
    today: date | None = None,
    rng: random.Random | None = None,
    reviewed_at: str | None = None,
) -> SrsState:
    """回傳評分後的新排程狀態（不修改傳入的 state）。"""
    validate_rating(rating)
    today = today or date.today()

    raw = next_interval(state, rating)
    if rating != AGAIN:
        raw = apply_jitter(raw, rng)
    interval = max(1.0, min(float(MAX_INTERVAL_DAYS), round(raw, 3)))

    ease = clamp_ease(state.ease_factor + EASE_DELTA[rating])
    lapses = state.lapse_count + (1 if rating == AGAIN else 0)

    return replace(
        state,
        interval=interval,
        ease_factor=ease,
        due_date=(today + timedelta(days=int(round(interval)))).isoformat(),
        review_count=state.review_count + 1,
        lapse_count=lapses,
        last_reviewed=reviewed_at or now_iso(),
    )


def new_state(card_id: int, track: str, today: date | None = None) -> SrsState:
    """建立一張卡在某個 track 上的初始狀態（今天到期）。"""
    today = today or date.today()
    return SrsState(
        card_id=card_id,
        track=track,
        interval=0.0,
        ease_factor=DEFAULT_EASE,
        due_date=today.isoformat(),
        review_count=0,
        lapse_count=0,
        last_reviewed=None,
    )


def describe_next(state: SrsState, today: date | None = None) -> str:
    """把下次到期時間講成人話，用在評分後的即時回饋。"""
    today = today or date.today()
    due = state.due
    if not due:
        return "下次：未排程"
    days = (due - today).days
    if days <= 0:
        return "下次：今天稍後"
    if days == 1:
        return "下次：明天"
    if days < 30:
        return f"下次：{days} 天後"
    months = days / 30.0
    return f"下次：約 {months:.1f} 個月後"


def maturity(state: SrsState) -> str:
    """卡片成熟度分級，統計用。"""
    if state.review_count == 0:
        return "new"
    if state.interval < 21:
        return "learning"
    return "mature"

File: js/srs.js
// 日期工具：全程用 YYYY-MM-DD 字串，避免時區把到期日算歪。
const Dates = (() => {
  function todayISO() {
    const d = new Date();
    return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}-${String(d.getDate()).padStart(2, '0')}`;
  }
  function nowISO() {
    const d = new Date();
    return `${todayISO()} ${String(d.getHours()).padStart(2, '0')}:${String(d.getMinutes()).padStart(2, '0')}`;
  }
  function parse(iso) {
    const [y, m, d] = String(iso).slice(0, 10).split('-').map(Number);
    return new Date(y, (m || 1) - 1, d || 1);
  }
  function format(date) {
    return `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, '0')}-${String(date.getDate()).padStart(2, '0')}`;
  }
  function addDays(iso, days) {
    const d = parse(iso);
    d.setDate(d.getDate() + days);
    return format(d);
  }
  function diffDays(isoA, isoB) {
    return Math.round((parse(isoA) - parse(isoB)) / 86400000);
  }
  return { todayISO, nowISO, parse, format, addDays, diffDays };
})();

// SM-2 間隔重複演算法 —— 與 cli/ielts/srs.py 同一組參數，兩邊行為一致。
// 純函式，不碰 DOM 也不碰 localStorage。
const SRS = (() => {
  const AGAIN = 1, HARD = 2, GOOD = 3, EASY = 4;
  const RATINGS = [AGAIN, HARD, GOOD, EASY];

  const RATING_LABELS = {
    1: 'Again 忘了',
    2: 'Hard 有點吃力',
    3: 'Good 想得起來',
    4: 'Easy 太簡單'
  };
  const RATING_SHORT = { 1: 'Again', 2: 'Hard', 3: 'Good', 4: 'Easy' };

  const MIN_EASE = 1.3, MAX_EASE = 3.0, DEFAULT_EASE = 2.5;
  const MAX_INTERVAL_DAYS = 365 * 2;
  const FIRST_INTERVAL = 1, SECOND_INTERVAL = 6, LAPSE_INTERVAL = 1;
  const HARD_MULTIPLIER = 1.2, EASY_BONUS = 1.3;
  const EASE_DELTA = { 1: -0.20, 2: -0.15, 3: 0, 4: 0.15 };
  const JITTER_MIN_INTERVAL = 4, JITTER_RATIO = 0.05;

  const clampEase = v => Math.max(MIN_EASE, Math.min(MAX_EASE, Math.round(v * 10000) / 10000));

  function newState() {
    return {
      interval: 0,
      ease: DEFAULT_EASE,
      due: Dates.todayISO(),
      reviews: 0,
      lapses: 0,
      lastReviewed: null
    };
  }

  function nextInterval(state, rating) {
    if (!RATINGS.includes(rating)) throw new Error('評分必須是 1-4');
    const interval = Math.max(0, state.interval || 0);
    const ease = clampEase(state.ease || DEFAULT_EASE);

    if (rating === AGAIN) return LAPSE_INTERVAL;
    if (state.reviews === 0 || interval <= 0) {
      return rating === EASY ? SECOND_INTERVAL : FIRST_INTERVAL;
    }
    if (interval < SECOND_INTERVAL && state.reviews === 1) {
      if (rating === HARD) return Math.max(FIRST_INTERVAL + 1, interval * HARD_MULTIPLIER);
      if (rating === EASY) return SECOND_INTERVAL * EASY_BONUS;
      return SECOND_INTERVAL;
    }
    if (rating === HARD) return interval * HARD_MULTIPLIER;
    if (rating === GOOD) return interval * ease;
    return interval * ease * EASY_BONUS;
  }

  // 4 天以上的間隔加 ±5% 抖動，讓到期日自然分散，不會全擠在同一天。
  function applyJitter(interval) {
    if (interval < JITTER_MIN_INTERVAL) return interval;
    return interval * (1 + (Math.random() * 2 - 1) * JITTER_RATIO);
  }

  function schedule(state, rating, today) {
    const base = state || newState();
    let raw = nextInterval(base, rating);
    if (rating !== AGAIN) raw = applyJitter(raw);
    const interval = Math.max(1, Math.min(MAX_INTERVAL_DAYS, Math.round(raw * 1000) / 1000));
    return {
      interval,
      ease: clampEase((base.ease || DEFAULT_EASE) + EASE_DELTA[rating]),
      due: Dates.addDays(today || Dates.todayISO(), Math.round(interval)),
      reviews: (base.reviews || 0) + 1,
      lapses: (base.lapses || 0) + (rating === AGAIN ? 1 : 0),
      lastReviewed: Dates.nowISO()
    };
  }

  function describeNext(state, today) {
    const days = Dates.diffDays(state.due, today || Dates.todayISO());
    if (days <= 0) return '下次：今天稍後';
    if (days === 1) return '下次：明天';
    if (days < 30) return `下次：${days} 天後`;
    return `下次：約 ${(days / 30).toFixed(1)} 個月後`;
  }

  function maturity(state) {
    if (!state || state.reviews === 0) return 'new';
    return state.interval < 21 ? 'learning' : 'mature';
  }

  return {
    AGAIN, HARD, GOOD, EASY, RATINGS, RATING_LABELS, RATING_SHORT,
    DEFAULT_EASE, MIN_EASE, MAX_INTERVAL_DAYS,
    newState, nextInterval, applyJitter, schedule, describeNext, maturity
  };
})();
